parsl/tests/test_flux.py

```diff
@@ -0,0 +1,170 @@
+import os
+import concurrent.futures as cf
+
+import pytest
+
+from parsl.app.errors import AppException
+from parsl.executors.flux.executor import (
+    FluxExecutor,
+    FluxFutureWrapper,
+    _complete_future,
+)
+
+try:
+    import flux.job.executor  # noqa: F401
+except ImportError:
+    FLUX_AVAIL = False
+else:
+    FLUX_AVAIL = True
+
+require_flux = pytest.mark.skipif(
```

Comment:
I'm not super fond of this style of magic autodetection of if a test should happen. Elsewhere we use explicit pytest `-k` arguments to disable tests we expect to not pass (see eg. `-k not cleannet and not issue363` in the Makefile). Every time I've encountered this style of testing, I've encountered tests incorrectly passing because this code will incorrectly not test flux when I've installed it incorrectly.

Reply:
That makes sense to me. There are three separate sets of tests in the module:

1. Tests that can run any time
2. Tests that require an installation of flux to be in `PATH`
3. Tests that must be run inside of a Flux instance, e.g. `flux start python -m pytest test_module.py`

3 implies 2. Should I split 1 and (2, 3) into separate modules, and then add a target "flux_test" for explicitly checking tests of type 2 and 3?

Also I'm a little confused about `-k`/`-m` because it looks like (according to pytest documentation) `-k` is for matching tests by name and `-m` is for matching tests by marker, and yet `-k` seems to be used in Parsl for matching tests by marker... am I missing something?

Reply:
huh, interesting. I hadn't noticed that -k vs -m thing.

I think you should get this PR merged (by clicking "update from master" and then waiting for a parsl committer to merge it, without putting in any further changes), and then fiddle with tests separately.

I don't think the layout of the test modules matters too much - just ensure the ones that require flux behaviour have a proper tag attached to them to make them disabled in the CI where necessary. Adding a flux test makefile target would be ok too but not mandatory.